types/ramda/index.d.ts

```diff
@@ -1027,6 +1028,13 @@ declare namespace R {
          */
         memoize<T = any>(fn: (...a: any[]) => T): (...a: any[]) => T;
 
+        /**
+         * A customisable version of R.memoize. memoizeWith takes an additional function that will be applied to a given
+         * argument set and used to create the cache key under which the results of the function to be memoized will be stored.
+         * Care must be taken when implementing key generation to avoid clashes that may overwrite previous entries erroneously.
+         */
+        memoizeWith<T extends (...args: any[]) => any>(keyFn: (...v: any[]) => string, fn: T): T;
```

Comment:
Is there any way to just parameterize `(...args: any[])` without the `=> any`? If there is, then perhaps we could get a narrower definition on this?

Reply:
Sorry, but I’m not sure what do you mean.

Reply:
With the current definition any function type will be accepted as `T`. So whatever function you pass to `fn`, the second parameter, you get a function with the parameters and the return value of that function out of `memoizeWith`. So there is no need for narrowing there (as far as I can tell).

There is of course the `keyFn` parameter, which would ideally be a function with the same parameters as `fn` but always a string return value. I'm not sure but it might be achievable with conditional types.

Reply:
To be honest, I don't want to force people to use `fn` and `keyFn` functions with the same arguments. Someone can easily need to memoize by, for example, just a first argument and forcing the same arguments for both functions can be quite annoying. 

Also, I think that the whole purpose of a `keyFn` is to return some string, nothing more. The fact that the Ramda passes a provided parameters to `keyFn` is just a convenient bonus. So, forcing the arguments is probably not a very good idea.

Reply:
if they don't accept the same argument lists though, then how would you reasonably be able to memoize stuff? I understand that maybe you don't want to use the full set of arguments, but that's a specific optimization, the function signatures have to accept the same set of arguments or it wouldn't actually be memoization. I actually can't think of any situation where the functions would accept different parameter sets and produce expected behavior from the function.

Reply:
Well, Ramda is basically a helper library. When you want to convice people to use it, it must be simple to use. You can already have a good function, which returns a "good enought" string. This function can easily require less params then function, which you want to memoize. By forcing the same arguments list, you need to introduce a wrapper, which is basically a boilerplate code, to make TypeScript happy. Then, is Ramda really helping you if you need to introduce boilerplate to make type checker happy? That is a reason why people moved to languages without typedefs - a lot of boilerplate just to satisfy compiler.

Also, we are still in JavaScript. Providing more parameters to function, which actually accepts less arguments is a language feature. We are using TypeScript to help us, not to change the language behaviour.

Consider lodash's `memoize` function. You can create the same memoize function in ramda by `memoizeWith(identity, fn)`. It will always use just a first argument.

On  the other side, I fully understand your suggestion. My opinion is just that a usage of helper library (Ramda) should be as easy as possible and without a necessary boilerplate.

Reply:
This is an argument about nothing.

TypeScript will happily allow a function that accepts fewer arguments than suggested by a type, so the argument about wrapper functions is misguided at best. And if `keyFn` *requires* more arguments than `fn`, this would indeed be a type error. Remember, TypeScript uses *structural* equivalence, not *nominal*, so if types *can* be reasonably unified, they *will* be.

However, there is currently no good way to express such relation between functions in TypeScript. There is a horrible way kinda similar to what we did in C++ before variadic templates, which I do not at all advocate using, and if anything I'm advocating against, but I am providing it here as a reference to make a point nonetheless:

- [TS Playground](https://www.typescriptlang.org/play/#src=type%20SubsString%3CT%20extends%20(...args%3A%20any%5B%5D)%20%3D%3E%20any%3E%20%3D%0AT%20extends%20(x1%3A%20infer%20U1%2C)%20%3D%3E%20any%20%3F%20(x1%3A%20U1%2C)%20%3D%3E%20string%3A%0AT%20extends%20(x1%3A%20infer%20U1%2Cx2%3A%20infer%20U2%2C)%20%3D%3E%20any%20%3F%20(x1%3A%20U1%2Cx2%3A%20U2%2C)%20%3D%3E%20string%3A%0AT%20extends%20(x1%3A%20infer%20U1%2Cx2%3A%20infer%20U2%2Cx3%3A%20infer%20U3%2C)%20%3D%3E%20any%20%3F%20(x1%3A%20U1%2Cx2%3A%20U2%2Cx3%3A%20U3%2C)%20%3D%3E%20string%3A%0AT%20extends%20(x1%3A%20infer%20U1%2Cx2%3A%20infer%20U2%2Cx3%3A%20infer%20U3%2Cx4%3A%20infer%20U4%2C)%20%3D%3E%20any%20%3F%20(x1%3A%20U1%2Cx2%3A%20U2%2Cx3%3A%20U3%2Cx4%3A%20U4%2C)%20%3D%3E%20string%3A%0AT%20extends%20(x1%3A%20infer%20U1%2Cx2%3A%20infer%20U2%2Cx3%3A%20infer%20U3%2Cx4%3A%20infer%20U4%2Cx5%3A%20infer%20U5%2C)%20%3D%3E%20any%20%3F%20(x1%3A%20U1%2Cx2%3A%20U2%2Cx3%3A%20U3%2Cx4%3A%20U4%2Cx5%3A%20U5%2C)%20%3D%3E%20string%3A%0AT%20extends%20(x1%3A%20infer%20U1%2Cx2%3A%20infer%20U2%2Cx3%3A%20infer%20U3%2Cx4%3A%20infer%20U4%2Cx5%3A%20infer%20U5%2Cx6%3A%20infer%20U6%2C)%20%3D%3E%20any%20%3F%20(x1%3A%20U1%2Cx2%3A%20U2%2Cx3%3A%20U3%2Cx4%3A%20U4%2Cx5%3A%20U5%2Cx6%3A%20U6%2C)%20%3D%3E%20string%3A%0AT%20extends%20(x1%3A%20infer%20U1%2Cx2%3A%20infer%20U2%2Cx3%3A%20infer%20U3%2Cx4%3A%20infer%20U4%2Cx5%3A%20infer%20U5%2Cx6%3A%20infer%20U6%2Cx7%3A%20infer%20U7%2C)%20%3D%3E%20any%20%3F%20(x1%3A%20U1%2Cx2%3A%20U2%2Cx3%3A%20U3%2Cx4%3A%20U4%2Cx5%3A%20U5%2Cx6%3A%20U6%2Cx7%3A%20U7%2C)%20%3D%3E%20string%3A%0AT%20extends%20(x1%3A%20infer%20U1%2Cx2%3A%20infer%20U2%2Cx3%3A%20infer%20U3%2Cx4%3A%20infer%20U4%2Cx5%3A%20infer%20U5%2Cx6%3A%20infer%20U6%2Cx7%3A%20infer%20U7%2Cx8%3A%20infer%20U8%2C)%20%3D%3E%20any%20%3F%20(x1%3A%20U1%2Cx2%3A%20U2%2Cx3%3A%20U3%2Cx4%3A%20U4%2Cx5%3A%20U5%2Cx6%3A%20U6%2Cx7%3A%20U7%2Cx8%3A%20U8%2C)%20%3D%3E%20string%3A%0AT%20extends%20(x1%3A%20infer%20U1%2Cx2%3A%20infer%20U2%2Cx3%3A%20infer%20U3%2Cx4%3A%20infer%20U4%2Cx5%3A%20infer%20U5%2Cx6%3A%20infer%20U6%2Cx7%3A%20infer%20U7%2Cx8%3A%20infer%20U8%2Cx9%3A%20infer%20U9%2C)%20%3D%3E%20any%20%3F%20(x1%3A%20U1%2Cx2%3A%20U2%2Cx3%3A%20U3%2Cx4%3A%20U4%2Cx5%3A%20U5%2Cx6%3A%20U6%2Cx7%3A%20U7%2Cx8%3A%20U8%2Cx9%3A%20U9%2C)%20%3D%3E%20string%3A%0AT%20extends%20(x1%3A%20infer%20U1%2Cx2%3A%20infer%20U2%2Cx3%3A%20infer%20U3%2Cx4%3A%20infer%20U4%2Cx5%3A%20infer%20U5%2Cx6%3A%20infer%20U6%2Cx7%3A%20infer%20U7%2Cx8%3A%20infer%20U8%2Cx9%3A%20infer%20U9%2Cx10%3A%20infer%20U10%2C)%20%3D%3E%20any%20%3F%20(x1%3A%20U1%2Cx2%3A%20U2%2Cx3%3A%20U3%2Cx4%3A%20U4%2Cx5%3A%20U5%2Cx6%3A%20U6%2Cx7%3A%20U7%2Cx8%3A%20U8%2Cx9%3A%20U9%2Cx10%3A%20U10%2C)%20%3D%3E%20string%3A%0A(...args%3A%20any%5B%5D)%20%3D%3E%20string%0A%0Adeclare%20function%20memoizeWith%3CT%20extends%20(...args%3A%20any%5B%5D)%20%3D%3E%20any%3E(keyFn%3A%20SubsString%3CT%3E%2C%20fn%3A%20T)%3A%20T%0A%0Aconst%20f%20%3D%20memoizeWith(()%20%3D%3E%20%22%22%2C%20(a%3A%20number%2C%20b%3A%20number%2C%20c%3A%20string)%20%3D%3E%20a%20%2B%20b%20%2B%20c)%20%2F%2F%20ok%0Aconst%20g%20%3D%20memoizeWith((x1%3A%20number)%20%3D%3E%20%22%22%2C%20(a%3A%20number%2C%20b%3A%20number%2C%20c%3A%20string)%20%3D%3E%20a%20%2B%20b%20%2B%20c)%20%2F%2F%20ok%0Aconst%20h%20%3D%20memoizeWith((x1%3A%20string)%20%3D%3E%20%22%22%2C%20(a%3A%20number%2C%20b%3A%20number%2C%20c%3A%20string)%20%3D%3E%20a%20%2B%20b%20%2B%20c)%20%2F%2F%20type%20error%0Aconst%20p%20%3D%20memoizeWith((x1%3A%20number%2C%20x2%3A%20number%2C%20x3%3A%20string)%20%3D%3E%20%22%22%2C%20(a%3A%20number%2C%20b%3A%20number%2C%20c%3A%20string)%20%3D%3E%20a%20%2B%20b%20%2B%20c)%20%2F%2F%20ok%0Aconst%20r%20%3D%20memoizeWith((x1%3A%20number%2C%20x2%3A%20number%2C%20x3%3A%20string%2C%20x4%3F%3A%20boolean)%20%3D%3E%20%22%22%2C%20(a%3A%20number%2C%20b%3A%20number%2C%20c%3A%20string)%20%3D%3E%20a%20%2B%20b%20%2B%20c)%20%2F%2F%20ok%0Aconst%20s%20%3D%20memoizeWith((x1%3A%20number%2C%20x2%3A%20number%2C%20x3%3A%20string%2C%20x4%3A%20boolean)%20%3D%3E%20%22%22%2C%20(a%3A%20number%2C%20b%3A%20number%2C%20c%3A%20string)%20%3D%3E%20a%20%2B%20b%20%2B%20c)%20%2F%2F%20type%20error)
- [Gist](https://gist.github.com/lierdakil/f4d1888dd1d708df0378a53740965812)

Reply:
@lierdakil Awesome, thanks for a comment.

I need to **apologies** to everyone as I didn't know that TypeScript will happily allows a function that accepts fewer arguments... It actually makes sense 🙂 

We should end this discussion with a conclusion:
We want to narrow `keyFn` types, but we need to wait (probably for an infinite time) as not everyone is using TS 2.8. So, lets merge it as it is now.